### ISSUE TITLE
Train and Test give different Recall scores

### ISSUE THREAD

**Issue by PaulHeraty** [ultralytics/yolov5#259]
Hi,
I'm training with the coco training and validation dataset. When I run training, I see the following test scores:

    Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    17/199     6.56G   0.07603     0.108   0.04793     0.232       226       320: 100%|██████████| 925/925 [04:13<00:00,  3.65it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 40/40 [00:17<00:00,  2.23it/s]
                 all       5e+03    3.63e+04       0.166       0.124      0.0648      0.0321

If I then take the last.pt checkpoint for that epoch and run the test script, I get similar Precision scores, but the Recall scores are always ~2x those reported by the train script.

               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|█████████████████████████████████████████████████| 40/40 [00:35<00:00,  1.12it/s]
                 all       5e+03    3.63e+04       0.172       0.252       0.129      0.0639

The command I use for training is:
python train --img 320 --batch 128 --epochs 200 --data ./data/coco.yaml --cfg ./models/yolov5s.yaml --cache

and for test it's:
python test.py --weights ./weights/last.pt --data ./data/coco.yaml --img 320 --batch 128

Why am I always seeing ~2x Recall scores in the test script, given that training calls that script also? It also looks like the runtime is ~2x longer in the standalone version. Is there some double counting going on?

I'm using the latest code as of today, but also saw this in the version I pulled on Jun 15th. I haven't changed any code and am using the data/model files without any modification.

Regards,
Paul

**Comment by glenn-jocher** [comment]
@PaulHeraty first of all, there's no point micromanaging your training so closely. Wait until it finishes.

Second of all, P and R are irrelevant, as they are user-setable to whatever value you want per your use case. mAP is the metric that matters. See https://github.com/ultralytics/yolov3/issues/898

**Comment by PaulHeraty** [comment]
Hi Glenn,

thanks for the quick reply. I don't usually look at P/R/AP during the run; I just used that epoch to highlight the issue. It's the very same at the end of the run, where Recall is ~2x larger when 'test' is run on it's own, as opposed to the test() is called from within train.

You can also see that mAP is ~2x different for me also, which is also true at the end of the run. So the main point is, the validation scores (mAP, mAP50 and Recall) are very different (~2x) between what train and test standalone report.

Regards,
Paul

**Comment by yxNONG** [comment]
@PaulHeraty 
did you use 2 gpu?

**Comment by glenn-jocher** [comment]
@PaulHeraty I would wait until training finishes, as early results are often unpredictable. COCO training typically produces mAPs within 1-2% of their training results when tested standalone. pycocotools always reports slightly higher mAP than the repo computation as well, maybe 1-2% higher.

In any case, running test.py directly will always provide you the most accurate metrics.

**Comment by yxNONG** [comment]
@glenn-jocher i believe test and train are different when use **n**-gpu(result in train is 1/**n** compare to test) in custom data

in train.py, call the `test.test(model=ema.ema)` may cause the different when use `distributeddataparallel `, however, `dataparallel` work fine , so i use it instead. (pytorch=1.4)

i modify `test.test(model=ema.ema)` to `test.test(model=model=ema.ema.module if hasattr(model, 'module') else ema.ema,)`
which is same while call test.py alone, it seems like it work for `distributeddataparallel `
  
already submit a new PR

**Comment by PaulHeraty** [comment]
@yxNONG yes, I'm using 2 GPUs.

**Comment by yxNONG** [comment]
@PaulHeraty that prove my guess
test.py report the true result.
i run with 4 gpu and got 1/4 by train.py compare to single gpu.
i guess in train.py it only count the result in 1 gpu. 

solution:
1.i submit a PR, it work for me, you can try to see whether it solve the problem for you or not.
2.if you like to use `model = torch.nn.DataParallel(model)`, 
   then replace the `model = torch.nn.parallel.DistributedDataParallel(model)` will also work, i use it for few week, the train               
   result reported is consist to test result
3.use single gpu

**Comment by PaulHeraty** [comment]
@yxNONG I made the changes from your PR to utils.py and train.py locally and re-ran. I now see very similar metrics reported for Recall and mAP when running on 2 GPUs.

Thanks!

**Comment by glenn-jocher** [comment]
@yxNONG @PaulHeraty ah, good detective work, thanks guys!!

To give some background on this, train.py used DDP and test.py DP by default until about 1 month ago, when we transitioned all models to FP16 storage and inference. During the change **we could not get FP16 inference to work on DP** in test.py, so we simply removed DP from test.py. Now test.py and detect.py are single-device FP16 inference. This change in test.py is likely the cause of the mAP drop now during multi-device training.

I'll take a look at the PR.
